game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleSteps.java

```diff
@@ -71,9 +71,12 @@
     final BattleStep airDefendVsNonSubs = new AirDefendVsNonSubsStep(this);
 
     final List<String> steps = new ArrayList<>();
-    steps.addAll(offensiveAaStep.getNames());
-
-    steps.addAll(defensiveAaStep.getNames());
+    if (offensiveAaStep.getOrder() != BattleStep.Order.NOT_APPLICABLE) {
+      steps.addAll(offensiveAaStep.getNames());
```

Comment:
I think there might be something of a feature envy code smell here. Notably we query an object for its state, then depending on that state we query the same object again for more state. The feature envy is on `BattleSteps` with respect to the state of `offensiveAaStep`. In practical sense this indicates a lack of data encapsulation.

This could potentially be fixed by moving the `getOrder()` concern to `offensiveAaStep`. In such a case there would be no conditional in `BattleSteps`. In other words, `offensiveAaStep` would evaluate its own order, if not applicable it would return an empty `getNames()` list, otherwise it would return some values.

WDYT?

Reply:
I agree with the code smell but that is because this is an intermediate state.  There are steps that will return different values for `getOrder` depending on their situation.  Once all of the steps are converted, I can grab the list of steps with code like:
```
List<BattleStep> steps = new ArrayList<>();
steps.add(new OffensiveAaFire(this, battleActions));
steps.add(new DefensiveAaFire(this, battleActions));
steps.add(new ...);
steps.add(new ...);

List<String> stepNames = steps.stream()
        .filter(step -> step.getOrder() != BattleStep.Order.NOT_APPLICABLE)
        .sorted(Comparator.comparing(BattleStep::getOrder))
        .flatMap(step -> step.getNames().stream())
        .collect(Collectors.toList());
```

Another idea, is that instead of allowing a step to have different positions in the battle order, I could just create duplicate steps.  They could inherit from the same parent class where all the main logic is and then override a method that determines whether `getName` or `executable` should be run.

Then the code to get the steps would look more like:
```
List<BattleStep> steps = new ArrayList<>();
steps.add(New OffensiveAaFire(this, battleActions));
steps.add(new SubRetreatBeforeBattle(this, battleActions));
steps.add(...<some other step>);
steps.add(new SubRetreatAfterBattle(this, battleActions));
steps.add(...<other steps>);

List<String> stepNames = steps.stream()
        .flatMap(step -> step.getNames().stream())
        .collect(Collectors.toList());
```

In that case, both `SubRetreatBeforeBattle` and `SubRetreatAfterBattle` would extend `SubRetreat` but the first class would check for `getSubRetreatBeforeBattle() == true` while the second class would check if `getSubRetreatBeforeBattle() == false`.

What would you prefer?  More classes or less classes?  Or maybe you have another idea on how to handle steps that don't have a static position in the order but can change depending on the state at the beginning of the round.

Reply:
@trevan 
I was thinking this:
```

List<String> stepNames = steps.stream()
        .sorted(Comparator.comparing(BattleStep::getOrder))
        .flatMap(step -> step.getNames().stream())
        .collect(Collectors.toList());
```

If you make BattleStep implement Comparable, perhaps could get even go simpler:

```
public interface BattleStep extends IExecutable, Comparable<BattleStep> {
 :
 :
  @Override
  default int compareTo(BattleStep rhs) {
    return Integer.compare(getOrder().ordinal(), rhs.getOrder().ordinal());
  }
```

```
List<String> stepNames = steps.stream()
        .sorted()
        .flatMap(step -> step.getNames().stream())
        .collect(Collectors.toList());
```

Reply:
@DanVanAtta so no pre filtering?  I'll try that out.  The existing code does do pre-filtering on the IExecutables that are added and I was trying to allow that with the `NOT_APPLICABLE` order but that might just be a relic.

Reply:
Basically the filtering would be internalized to each BattleStep. Each one would be responsible to know if it applies and if so generating any necessary step names.

Reply:
@DanVanAtta , I'm not just talking about step names but the executable part as well.  I'm worried about the lines such as:

https://github.com/triplea-game/triplea/blob/f524e9819a292c6c3bfde535c7f4d90bb94088ca/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L1105
https://github.com/triplea-game/triplea/blob/f524e9819a292c6c3bfde535c7f4d90bb94088ca/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L1118
https://github.com/triplea-game/triplea/blob/f524e9819a292c6c3bfde535c7f4d90bb94088ca/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L1131
https://github.com/triplea-game/triplea/blob/f524e9819a292c6c3bfde535c7f4d90bb94088ca/game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java#L1475

I'm not entirely sure why the check for those steps is outside of the executable.  A check outside of the executable happens at the start of the round while the checks inside of the executable happen when the step actually happens.  So the checks are being done on different data.  Most of the executables have all their checks inside of them or the checks outside of them are just determining the order of the step.

It might be that I can move all of those checks inside of the executable so that they run when the step occurrs but I'm not positive at this moment in time and so the `getOrder() != NOT_APPLICABLE` was mainly for those steps.  I guess I can remove the NOT_APPLICABLE order and when I get to converting those steps, figure out if I need NOT_APPLICABLE or not.

Reply:
Oh, in addition, by not adding the step to the stack, they don't get saved in saved games.  That is another thing that I have to look into when I convert them.  By not having the `getOrder() != NOT_APPLICABLE` check, all of the steps will be added to the saved game, including ones that were previously not added.

Reply:
@DanVanAtta I've made the change.  I've also tested a scenario that would originally have had just offensive AA steps in the save and no defensive AA steps in the save.  I created a save with master and with this branch and then loaded both saves in this branch.  It loaded fine and it looks like it played out the rest of the round without issues.

Reply:
Interesting context, thank you. Sounds as though we should err on the side of caution.